GVFS/GVFS.UnitTests/Common/NuGetUpgraderTests.cs

```diff
@@ -0,0 +1,300 @@
+
+            success.ShouldBeTrue($"Expecting TryQueryNewestVersion to have completed sucessfully. Error: {message}");
+            newVersion.ShouldNotBeNull();
+
+            bool downloadSuccessful = this.upgrader.TryDownloadNewestVersion(out message);
+            downloadSuccessful.ShouldBeFalse();
+        }
+
+        [TestCase]
+        public void AttemptingToDownloadBeforeQueryingFails()
+        {
+            string message;
+            List<IPackageSearchMetadata> availablePackages = new List<IPackageSearchMetadata>()
+            {
+                this.GeneratePackageSeachMetadata(new Version(CurrentVersion)),
+                this.GeneratePackageSeachMetadata(new Version(NewerVersion)),
+            };
+
+            IPackageSearchMetadata newestAvailableVersion = availablePackages.Last();
+
+            string downloadPath = "c:\\test_download_path";
+            this.mockNuGetFeed.Setup(foo => foo.QueryFeedAsync(NuGetFeedName)).ReturnsAsync(new List<IPackageSearchMetadata>(availablePackages));
+            this.mockNuGetFeed.Setup(foo => foo.DownloadPackage(It.Is<PackageIdentity>(packageIdentity => packageIdentity == newestAvailableVersion.Identity))).Returns(Task.FromResult(downloadPath));
+
+            bool downloadSuccessful = this.upgrader.TryDownloadNewestVersion(out message);
+            downloadSuccessful.ShouldBeFalse();
+        }
+
+        [TestCase]
+        public void TestUpgradeAllowed()
+        {
+            // Properly Configured NuGet config FeedUrlForCredentials
+            NuGetUpgrader.NugetUpgraderConfig nuGetUpgraderConfig =
+                new NuGetUpgrader.NugetUpgraderConfig(this.tracer, null, NuGetFeedUrl, NuGetFeedName, NuGetFeedUrlForCredentials);
+
+            NuGetUpgrader nuGetUpgrader = new NuGetUpgrader(
+                CurrentVersion,
+                this.tracer,
+                nuGetUpgraderConfig,
+                false,
+                this.mockFileSystem.Object,
+                this.mockNuGetFeed.Object,
+                new LocalUpgraderServices(this.tracer, this.mockFileSystem.Object));
+
+            nuGetUpgrader.UpgradeAllowed(out string _).ShouldBeTrue("NuGetUpgrader config is complete: upgrade should be allowed.");
+
+            // Empty FeedURL
+            nuGetUpgraderConfig =
+                new NuGetUpgrader.NugetUpgraderConfig(this.tracer, null, string.Empty, NuGetFeedName, NuGetFeedUrlForCredentials);
+
+             nuGetUpgrader = new NuGetUpgrader(
+                CurrentVersion,
+                this.tracer,
+                nuGetUpgraderConfig,
+                false,
+                this.mockFileSystem.Object,
+                this.mockNuGetFeed.Object,
+                new LocalUpgraderServices(this.tracer, this.mockFileSystem.Object));
+
+            nuGetUpgrader.UpgradeAllowed(out string _).ShouldBeFalse("Upgrade without FeedURL configured should not be allowed.");
```

Comment:
I'm surprised our StyleCop rules allow discards (`out string _`). Does it also permit type-less discards too?
I.e., just `out _`.